File: routes/produtos.js
const express = require('express');
const { PrismaClient } = require('@prisma/client');
const { successResponse, errorResponse } = require('../utils/responses');
const { authenticateToken, requireAdmin } = require('../middleware/auth');

const router = express.Router();
const prisma = new PrismaClient();

// Get produtos
router.get('/', async (req, res) => {
  try {
    const { categoria, search, promocao, skip = 0, limit = 50 } = req.query;

    const where = { is_active: true };

    if (categoria) {
      where.categoria = categoria;
    }

    if (search) {
      where.OR = [
        { nome: { contains: search, mode: 'insensitive' } },
        { descricao: { contains: search, mode: 'insensitive' } }
      ];
    }

    if (promocao !== undefined) {
      where.promocao = promocao === 'true';
    }

    const produtos = await prisma.product.findMany({
      where,
      skip: parseInt(skip),
      take: parseInt(limit),
      orderBy: { createdAt: 'desc' }
    });

    return successResponse(res, produtos, 'Produtos listados com sucesso');

  } catch (error) {
    console.error(error);
    return errorResponse(res, 'Erro interno do servidor', 500);
  }
});

// Get produto by ID
router.get('/:id', async (req, res) => {
  try {
    const { id } = req.params;

    const produto = await prisma.product.findFirst({
      where: {
        id: parseInt(id),
        is_active: true
      }
    });

    if (!produto) {
      return errorResponse(res, 'Produto não encontrado', 404);
    }

    return successResponse(res, produto, 'Produto encontrado');

  } catch (error) {
    console.error(error);
    return errorResponse(res, 'Erro interno do servidor', 500);
  }
});

// Create produto (admin only)
router.post('/', authenticateToken, requireAdmin, async (req, res) => {
  try {
    const produto = await prisma.product.create({
      data: req.body
    });

    return successResponse(res, produto, 'Produto criado com sucesso', 201);

  } catch (error) {
    console.error(error);
    return errorResponse(res, 'Erro interno do servidor', 500);
  }
});

// Update produto (admin only)
router.put('/:id', authenticateToken, requireAdmin, async (req, res) => {
  try {
    const { id } = req.params;

    const produto = await prisma.product.update({
      where: { id: parseInt(id) },
      data: req.body
    });

    return successResponse(res, produto, 'Produto atualizado com sucesso');

  } catch (error) {
    console.error(error);
    return errorResponse(res, 'Produto não encontrado', 404);
  }
});

// Delete produto (admin only)
router.delete('/:id', authenticateToken, requireAdmin, async (req, res) => {
  try {
    const { id } = req.params;

    await prisma.product.update({
      where: { id: parseInt(id) },
      data: { is_active: false }
    });

    return successResponse(res, null, 'Produto deletado com sucesso');

  } catch (error) {
    console.error(error);
    return errorResponse(res, 'Produto não encontrado', 404);
  }
});

module.exports = router;

File: app/routers/produtos.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from sqlalchemy import or_
from pydantic import BaseModel
from typing import List, Optional
from app.database import get_db
from app.models.product import Product
from app.auth import require_admin
from app.utils import success_response, error_response

router = APIRouter(prefix="/produtos", tags=["Produtos"])

class ProductCreate(BaseModel):
    nome: str
    descricao: str = None
    preco: float
    imagens: List[str] = []
    categoria: str
    promocao: bool = False
    preco_promocional: float = None
    estoque: int = 0

class ProductUpdate(BaseModel):
    nome: str = None
    descricao: str = None
    preco: float = None
    imagens: List[str] = None
    categoria: str = None
    promocao: bool = None
    preco_promocional: float = None
    estoque: int = None

@router.get("/")
def get_produtos(
    categoria: Optional[str] = None,
    search: Optional[str] = None,
    promocao: Optional[bool] = None,
    skip: int = 0,
    limit: int = 50,
    db: Session = Depends(get_db)
):
    query = db.query(Product).filter(Product.is_active == True)
    
    if categoria:
        query = query.filter(Product.categoria == categoria)
    
    if search:
        query = query.filter(or_(
            Product.nome.ilike(f"%{search}%"),
            Product.descricao.ilike(f"%{search}%")
        ))
    
    if promocao is not None:
        query = query.filter(Product.promocao == promocao)
    
    products = query.offset(skip).limit(limit).all()
    return success_response(data=products, message="Produtos listados com sucesso")

@router.get("/{product_id}")
def get_produto(product_id: int, db: Session = Depends(get_db)):
    product = db.query(Product).filter(Product.id == product_id, Product.is_active == True).first()
    if not product:
        return error_response("Produto não encontrado", 404)
    return success_response(data=product, message="Produto encontrado")

@router.post("/", dependencies=[Depends(require_admin)])
def create_produto(product_data: ProductCreate, db: Session = Depends(get_db)):
    product = Product(**product_data.dict())
    db.add(product)
    db.commit()
    db.refresh(product)
    return success_response(data=product, message="Produto criado com sucesso")

@router.put("/{product_id}", dependencies=[Depends(require_admin)])
def update_produto(product_id: int, product_data: ProductUpdate, db: Session = Depends(get_db)):
    product = db.query(Product).filter(Product.id == product_id).first()
    if not product:
        return error_response("Produto não encontrado", 404)
    
    for field, value in product_data.dict(exclude_unset=True).items():
        setattr(product, field, value)
    
    db.commit()
    return success_response(data=product, message="Produto atualizado com sucesso")

@router.delete("/{product_id}", dependencies=[Depends(require_admin)])
def delete_produto(product_id: int, db: Session = Depends(get_db)):
    product = db.query(Product).filter(Product.id == product_id).first()
    if not product:
        return error_response("Produto não encontrado", 404)
    
    product.is_active = False
    db.commit()
    return success_response(message="Produto deletado com sucesso")